controllers/tf_controller.go

```diff
@@ -295,6 +295,28 @@ func (r *TerraformReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		r.RecordReadiness(ctx, &terraform)
 	}
 
+	// Reset retry count if necessary.
+	revisionChanged := sourceObj.GetArtifact().Revision != terraform.Status.LastAttemptedRevision
+	generationChanges := terraform.Generation != terraform.Status.ObservedGeneration
+	if revisionChanged || generationChanges {
+		log.Info("Reset reconciliation failures count. Reason: resource changed")
+		terraform.ResetReconciliationFailures()
+		if err := r.patchStatus(ctx, req.NamespacedName, terraform.Status); err != nil {
+			log.Error(err, "unable to update status after planing")
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	if !terraform.ShouldRetry() {
+		log.Info(fmt.Sprintf(
+			"Resource reached maximum number of retries (%d/%d). Generation: %d",
+			terraform.GetReconciliationFailures(),
+			terraform.Spec.Remediation.Retries,
```

Comment:
True, originally it was in an extra if block.

// TODO

Reply:
It's not a NPR.
`ShouldRetry()` will return true if `in.Spec.Remediation == nil`, because of that, if `!ShouldRetry()` is true, the value is not `nil`.

```go
func (in *Terraform) ShouldRetry() bool {
	if in.Spec.Remediation == nil {
		return true
	}
	// ...
```

I'll add an extra check just in case, but now it's not NPR.

Reply:
Added a comment about it to make is clear, but I feel testing it redundant.

Reply:
You are right, I got myself in a twist there. Best not to rely on it, though.